README.Docker.md

```diff
@@ -0,0 +1,246 @@
+but sharing the socket will equally work. While `clamdtop` works well as an
+example here, it is of course important to realize, this can also be used to
+connect a mail server to `clamd`.
+
+
+### TCP/UDP
+While `clamd` and `milterd` will listen on the default tcp ports as per
+configuration directives, docker does not expose these by default to the host.
+Only within  containers can these ports be accessed. To expose, or publish,
+these ports to the host, and thus potentially over the (inter)network, the
+`--publish` (or `--publish-all`) flag to `docker run` can be used.
+While more advanced/secure mappings can be done as per documentation, the basic
+way is to `--publish [<host_port>:]<container_port>` to make the port available
+to the host.
+```console
+       --publish 73310:3310 \
+       --publish 7357
+```
+The above would thus publish the milter port 3310 as 73310 on the host and the
+clamd port 7357 as a random to the host. The random port can be inspected via
+`docker ps`.
+
+**Warning** extreme caution is to be taken when using `clamd` over TCP/UDP as
+there are no protections on that level. All traffic is un-encrypted. Extra care
+is to be taken when using TCP/UDP communications.
+
+
+## Virus database
+The virus database in `/usr/share/clamav` is by default unique to each container
+and thus is normally not shared. For simple setups this is fine, where only one
+instance of `clamd` is expected to run in a dockerized environment. However
+some use cases may want to efficiently share the database. To do so, the
+`docker volume` command can be used to create a persistent virus database.
+
+
+## Container clamd healthcheck
+Docker has the ability to run simple `ping` checks on services running inside
+containers. If clamd is running inside the container, docker will occasionally
+send a `ping` to clamd on the default port and wait for the pong from clamd.
+If clamd fails to respond, docker will treat this as an error. The healthcheck
+results can be viewed with `docker inspect`.
+
+
+## Performance
+The performance impact of running `clamd` in docker is negligible. Docker is
+in essence just a wrapper around Linux's cgroups and cgroups can be thought of
+as `chroot` or FreeBSD's a `jail`. All code is executed on the host without
+any translation. Docker does however do some isolation (through cgroups) to
+isolate the various systems somewhat.
+
+Of course, nothing in life is free, and so there is some overhead. Disk-space
+being the most prominent one. The docker container might have some duplication
+of files for example between the host and the container. Further more, also RAM
+memory may be duplicated for each instance, as there is no RAM-deduplication.
+Both of which can be solved on the host however. A filesystem that supports
+disk-deduplication for exaple and a memory manager that does RAM-deduplication.
+
+But since the container in itself is already very small ~16 MiB at the time of
```

Comment:
This will be a bit bigger after we add in the databases.

Reply:
Agreed. I'll make a note, and instruct how to use the separate container layers when I do the work

Reply:
Wouldn't everyone who uses the container require the database?  Yes, it does push the problem from Cisco to Docker -- though if the users cache the images locally, then 
A) starting a new container from the cached image which contains the database won't require a download except to do the incremental update (a few very small downloads), while 
B) starting a new container from a cached image lacking the database would require a couple very large downloads _every time_.

Reply:
Actually, not at all. (Of course they do) but imagine that an org uses a centralized clam database volume/server, but spawns hundreds of clamav instances, because they want to be good netizens and save bandwidth. In that case, they don't want to pull the bigger container. (To be fair, they might be proxying the container then as well :p)

I do favor built-in solution though; it saves a lot of bandwith for the non-conscious passer by.

I think the proper 'solution' is to have various different tags. 'base' without database, incremental with. Since you want to regularly push a container (say once a week to get the latest database? (or once a month, that would be up to you of course) So you'd have `clamav:v1.2.3-base` and `clamav:v1.2.3` where `clamav:latest` would point to `clamav-v1.2.3` So users that just do `docker pull clamav` get latest, which includes the database. Maybe tagging `clamav:latest-base` would be fair to have, so people can track the latest. Point being, it requires extra work and extra thought to pull the image without the database, so you likely only get those pulls from users that actually know what they are doing. (The defaults should be so that 99% of the people do 'the right thing' (tm) right? :) anyhow)

One bigger problem is, due to my lacking of knowledge, is how to trigger an updated build :) Docker autobuilds works great due to it being triggered by git commits. This would be a forced trigger (API for example or docker build + docker push), which would be okay to do through a github action too, but even there, you need a trigger. ('run pipeline' button or a `make docker-update` could do that trick and i see the `make docker-update` trick.

In that case, you could let docker autobuilds do the regular builds that are triggerd because of git changes (which includes both the base and the database included builds) and manually push 'weekly database updates'. I'm not too familar with github actions, but I know that in GitLab for example, you'd manually trigger the pipeline via a 'run pipeline' button, which would do this (as it if there was a git commit) and you can even tell GitLab pipelines to run scheduled tasks (cron-jobs) which would do exactly what would be needed. Push new database container to docker hub.

(This is a bigger response that probably should have gone elsewhere to properly discuss this :)

Reply:
Your tag ideas make sense to me, and I like the `-base` suffix for the version w/out a database. Pushing a new image once a month would probably suffice. Our CDN keeps the last 90 database patch files, and they're quite small. 

We'll have to think about how to trigger updated builds.  Github Actions also has scheduled actions (https://docs.github.com/en/actions/reference/events-that-trigger-workflows#scheduled-events)
Another option would be to trigger these builds using a scheduled Jenkins job on our internal Jenkins server.

Reply:
I've been meaning to ask (and was to lazy to search I admit)
Thinking about this a bit over time, you could, in theory, just make this a job for the virus DB repo. I don't know where it is, how it works and how it's managed, but you could include a trigger to update the database everytime a 'big' update is done. e.g. not for daily changes, but the other ones. Then you have a predictable hook that is aligned with 'changes to the big database' which is ultimately what you'd want anyway.
You probably want to do this regardless, because you want to update the database in ALL your published containers, not just the 'latest' or master, but all tags you have/support with the database. TBC :)

Reply:
Our virus db build system is owned by another team here.  For me to get anything automated, the easiest way would be for me to write something to monitor the database and trigger docker builds if needed.  It'd probably be worthwhile to just put a time limit on it and refresh the images with the latest signatures every 2 months.  The daily patch files are very small.  If we do a large migration, the other team will give me a heads up and I can kick it off manually. 

We tend to only do security patches for the current minor version and sometimes the previous minor version if there was a recent release, so as to provide some overlap.  For example, we're no longer supporting 0.102.X as of today, and are asking everyone to upgrade to 0.103.2 (just published).

Reply:
Ok, as long as freshclam can deal with this, I guess it's fine.

I added another commit to this beastly MR :) which allows you to do just that, update the container and push the tag. I only tested it locally with my namespace (it works quite well).

What it does/how it works, it gets all tags from the requested repository on the requested registry. It then checks if for the postfix `_base` and for all those tags, pulls them, adds freshclam, and pushes them without the `_base` postfix. I was quite proud of that one :D in any case, you can very easily do this manually, scheduled or whatever else.

It would be the best, and nicest thing, for this script to live and triggered as I said, in the virus database pipeline. So you could then do this completely automated on big changes. I think that would be really sexy and very cool. Regardless, you can have it here too, and just run it manually. Plugging it into a .github action is deff. a really nice trick for this as well (just need to deal with credential storage in a reliable way) but I'm probably not going to work on it soon (no experience with github actions, nor there secret handling. I could have cooked up something super fast on gitlab CI :p
Regardless, I think it's probably super trivial. Maybe 1 or 2 lines of code. While the script can do the docker login stuff, if you can do it more securely 'outside' of it (e.g. docker login, run-script) then all you do is not passing a username/password and it will rely on the external login, if that makes sense.